rfcs/0007-handling-corruption.md

```diff
@@ -217,108 +209,184 @@ durability and availability systems on top of SlateDB.
+
+To ensure durability by default, I propose that `open_with_opts()` perform
+WAL recovery synchronously and return any WAL corruption found 
+as errors, aborting the database open process until the corruption is resolved.
+The error returned should be a distinct type that can easily be matched to 
+determine if the error is due to non-transient corruption or a transient error
+(such as network connectivity issues).
+
+#### Handling SST Corruption During Compaction
+Compaction is an asynchronous process that may encounter corrupted SST files
+during operation. If corruption is detected, the compaction process defers
+control to the `DBOptions.on_corruption` callback.
+
+- If the `DBOptions.on_corruption` callback returns an error, the compaction
+  operation aborts the current run.
+- If the user-defined `DBOptions.on_corruption` callback does not return an
+  error, compaction will ignore the corruption to the best of its ability,
+  which may result in data loss.
+
+Compaction will continue to attempt future runs, aborting each time corruption
+is encountered until the corruption is repaired or compaction is halted. If
+compaction were to never run again, it is possible that the operator might
+remain unaware that compaction has stopped and fail to restart it after
+repairing the corruption.
+
+#### Handling SST Corruption During Get
+If SlateDB detects corruption while searching for a key during a `get` call,
+the `DBOptions.on_corruption` callback is invoked.
+
+- If the callback returns an error, the `get` caller will receive an error that
+  includes the error returned by the callback.
+- If the callback does not return an error, no error will be returned to the
+  `get` caller.
+
+The error returned should be of a distinct type, allowing it to be easily
+identified as either non-transient corruption or a transient error (such as
+network connectivity issues).
+
+Since `get` calls may be ongoing during corruption, it is possible that the
+callback will be called continuously and by multiple threads. Therefore, any
+user-defined callback must be thread-safe. Additionally, the callback should
+avoid slow or blocking operations to ensure the database remains available if
+corruption is encountered.
+
+To support operators who value durability, the developer making the `get` call
+can choose to abort the application if a corruption error is returned. The
+application can then provide operators with detailed information included in
+the error returned by SlateDB, including the nature of the corruption and
+suggestions on how to fix it.
+
+##### Remaining Available
+In the case where corruption is encountered, and `DBOptions.on_corruption` allows
+operation to continue, SlateDB should endeavor to return the most recent 
+valid data it can.
+
+- If the SST Bloom Filter indicates a value exists in the SST, yet the key was
+  not found, and corruption was detected during iteration, the `get` caller
+  should receive a Corruption error.
```

Comment:
I don't totally follow this one. Should the caller get a Corruption error, or should on_corruption be invoked?

Reply:
This section was my attempt at calling out exceptions when `DBOptions.on_corruption` does not return an error, yet we know with near certainty that the key was involved in the corruption.

However, if we want to maintain semantic simplicity for `DBOptions.on_corruption` and avoid confusion, I think we should remove this section.

Reply:
Can you elaborate on this scenario? The thing that's jamming me up is that bloom filters can give false positives, so the bloom filter can't "indicate a value xists in the sst"--only that it probably exists. If "corruption was detected during iteration" wouldn't we call on_corruption no matter what?